### DISCUSSION TITLE
Issue in executing 'convert' shell command using PHP system() function

### DISCUSSION THREAD

**Discussion by armdeveloper99** [ImageMagick/ImageMagick#5999] [Help]
### ImageMagick version

 6.9.10-68

### Operating system, version and so on

Linux server

### Description

I have a statement in my PHP code for creating an image like

system('convert -depth '.$this->imagedepth.' -density '.$this->dpi.' -units pixelsperinch -background transparent -gravity '.$this->gravity.' -fill "#'.$this->fore.'" -font '.$this->font.' -stroke none -pointsize '.$this->size.' label:"'.$this->msg.'" '.WWW_ROOT.IMAGES_URL.'editorimages/'.$pngName);

This command is not creating any image file in the server. Also, I tried to execute the same shell command via the command line interface, and result in the following error

convert: unable to access configure file `colors.xml' @ warning/configure.c/GetConfigureOptions/677.

Any idea where I am going wrong?

**Comment by fmw42** [comment]
Looks like your install is missing a file.  Please show your result from 

`convert -version`

Also please post your command with real data rather than variables.  

Some systems do not allow the use of "system()" calls. Are you sure that works.  Can you use exec() in place of system()

**Comment by armdeveloper99** [comment]
Version: ImageMagick 6.9.10-68 Q16 x86_64 2021-10-14 https://imagemagick.org Copyright: © 1999-2019 ImageMagick Studio LLC License: https://imagemagick.org/script/license.php Features: Cipher DPC Modules OpenMP(3.1) Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib Delegates (built-in): bzlib cairo fontconfig freetype gslib jng jp2 jpeg lcms ltdl lzma openexr pangocairo png ps rsvg tiff wmf x xml zlib

convert -depth 64 -density 300 -units pixelsperinch -background transparent -gravity west -fill "#000000" -font /home4/iroof/public_html/shop.iroof.com/app/webroot/img/editorfonts/times.ttf -stroke none -pointsize 5 label:"dfdfdfd" /home4/iroof/public_html/shop.iroof.com/app/webroot/img/editorimages/16746294668602.png

Yes, system() command is working. The statement echo system('convert -version') works. exec() is working but doesn't execute the required statement.


Also facing the following issues while executing sample convert commands,
 
convert -size 320x85 canvas:none -font Bookman-DemiItalic -pointsize 72 \ -draw "text 25,60 \'Magick\'" -channel RGBA -blur 0x6 -fill darkred -stroke magenta \ -draw "text 20,55 \'Magick\'" /home4/iroof/public_html/shop.iroof.com/app/webroot/img/editorimages/fuzzy-magick.png

convert: unable to access configure file `colors.xml' @ warning/configure.c/GetConfigureOptions/677.
convert: unable to open image ` -draw': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564.
convert: unable to open image `text 25,60 \'Magick\'': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: unable to open image `text 25,60 \'Magick\'': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564.
convert: unable to open image ` -draw': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564.
convert: unable to open image `text 20,55 \'Magick\'': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: unable to open image `text 20,55 \'Magick\'': No such file or directory @ error/blob.c/OpenBlob/2882.
convert: no decode delegate for this image format `' @ error/constitute.c/ReadImage/564.

**Comment by snibgo** [comment]
You seem to have two problems: (1) unable to access colors.xml, and (2) a nested quoting issue.

On (1), I suspect your IM has not been properly installed. IM should be able to find its own XML files. I suggest re-installing IM.

On (2), I suggest trying the command at the command line. Remove the isolated backslashes "\\" so it looks like this:
```
convert -size 320x85 canvas:none -font Bookman-DemiItalic -pointsize 72 -draw "text 25,60 'Magick'" -channel RGBA -blur 0x6 -fill darkred -stroke magenta -draw "text 20,55 'Magick'" /home4/iroof/public_html/shop.iroof.com/app/webroot/img/editorimages/fuzzy-magick.png
```
Does that work? If not, then what is the error message?

Then you need to use that in a `system()` in PHP. This means passing the entire string, including the nested quotes. I don't use PHP and can't advise, but I suppose some nested quotes need to be escaped, so PHP doesn't think they signal the end of the entire string.

**Reply by armdeveloper99** [reply]
@snibgo with minor modifications I have tried executing the command 

convert -size 320x85 canvas:none -font DejaVu-Sans -pointsize 72 -draw "text 25,60 Magick" -channel RGBA -blur 0x6 -fill darkred -stroke magenta -draw "text 20,55 Magick" dddddd.png

this statement got executed but it returns the following error.

convert: unable to access configure file `colors.xml' @ warning/configure.c/GetConfigureOptions/677.

convert: unable to access configure file `type.xml' @ warning/configure.c/GetConfigureOptions/677.

convert: unable to access configure file `type.xml' @ warning/configure.c/GetConfigureOptions/677.

convert: unable to access configure file `type.xml' @ warning/configure.c/GetConfigureOptions/677.

The above command is executed in the terminal but is not get executed using the PHP system() function. system() function is executing other general commands like 'convert -version'.

**Comment by snibgo** [comment]
IM needs colors.xml to translate `magenta`, and type.xml to translate `DejaVu-Sans`. These are failing. I suggest reinstalling IM.

When the command works at the terminal, then tackle the PHP issue.

**Reply by armdeveloper99** [reply]
Do we need to mind the version of ImageMagick while installing it? The currently installed version is ImageMagick 6.9.10-68

The same functionality is working in other servers where the ImageMagick versions are 6.9.12-70 and 6.9.0-0

**Comment by snibgo** [comment]
> Do we need to mind the version of ImageMagick while installing it?

That is a systems management issue, and it is difficult to advise without knowing the full circumstances.

In general, I suggest:

- For new work, use the latest IM v7.

- When you have IM running on multiple machines, management is simpler if they are all the same version.

- When upgrading to a new version of IM, the upgrade path should include regression testing, eg execute a test suite of commands and ensure the results are as expected.

- If you have many scripts that were written for v6, upgrading to v7 may not be simple. If you stick to v6, you won't get the benefit of new features on v7. Any new work (or maintenance work on old scripts) should be written so they will work correctly on either v6 or v7.

**Comment by fmw42** [comment]
PHP may have its own version of Imagemagick independent of your installation.  Either way, you might try adding the full path to convert (or magick) depending upon the version and command you use in your PHP system command. If on IM 7, then use magick, not convert.
